### PR TITLE
O3-872: RefApp 3.x Patient Search E2E to run in a dockerized environment

### PR DESCRIPTION
**Pull request by kumuditha-udayanga**
### Purpose
The purpose of this PR is to fix [O3-872](https://issues.openmrs.org/browse/O3-872).

### Approach
- Run 3.x patient search E2E test in a dockerized environment

### CODE REVIEW

**Comment by jayasanka-sack** [comment]
I enabled the github action right now, Could you please push it again?

File: qaframework-bdd-tests/src/test/resources/features/refapp-3.x/06-patient-search/patient-search.feature [hunk 1]
```diff
@@ -2,11 +2,8 @@ Feature: Patient Search
...
-    Then the result should be "<result>"
-    Examples:
-      | patientName | result           |
-      | Kevin Jones | Found 1 patient  |
-      | 100MQ       | No results found |
+  @patient-search
+  Scenario: Search for a patient
+    When the user search for "John Doe"
+    Then the result should be "Found 1 patient"
+
```
Comment (jayasanka-sack) [inline qaframework-bdd-tests/src/test/resources/features/refapp-3.x/06-patient-search/patient-search.feature:9]:
```suggestion

```

File: qaframework-bdd-tests/docker/docker-compose-refqa-3x.yml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      MYSQL_DATABASE: openmrs
+      MYSQL_USER: openmrs
+      MYSQL_PASSWORD: Admin123
+      MYSQL_ROOT_PASSWORD: Admin123
+    volumes:
+      - db-data:/var/lib/mysql
+
+volumes:
+  openmrs-data: ~
+  db-data: ~
```
Comment (jayasanka-sack) [inline qaframework-bdd-tests/docker/docker-compose-refqa-3x.yml:57]:
```suggestion
  db-data: ~
 
```

File: .github/workflows/refapp-3x-patient-search.yml [hunk 3]
```diff
@@ -14,7 +14,15 @@ jobs:
...
-      - uses: actions/checkout@v2
+      - name: Checkout qaframework
+        uses: actions/checkout@master
+        with:
+          repository: ${{github.repository}}
+      - name: Run db and web containers
+        run: |
+          docker-compose -f docker/docker-compose-refqa-3x.yml up -d
+      - name: Wait for Openmrs instance to start
+        run: while [[ "$(curl -s -o /dev/null -w ''%{http_code}'' http://localhost/openmrs/spa/login)" != "200" ]]; do echo "$(curl -s -o /dev/null -w ''%{http_code}'' http://localhost/openmrs/spa/login)"; sleep 10; done
```
Comment (jayasanka-sack) [inline .github/workflows/refapp-3x-patient-search.yml:25]: No need to echo the status code.
```suggestion
      - name: Wait for Openmrs instance to start
        run: while [[ "$(curl -s -o /dev/null -w ''%{http_code}'' http://localhost/openmrs/spa/login)" != "200" ]]; sleep 10; done
```

Comment (jayasanka-sack) [reply .github/workflows/refapp-3x-patient-search.yml:25]: Also, we can reduce the sleep time to 2 - 3 seconds.

**Review by jayasanka-sack — CHANGES_REQUESTED** [review]
Thanks for the PR @kumuditha-udayanga! This will save a lot of time. I added some minor comments. Let's implement the same for other tests once @brandones and @ibacher review this.

File: qaframework-bdd-tests/src/test/resources/features/refapp-3.x/06-patient-search/patient-search.feature [hunk 4]
```diff
@@ -2,11 +2,7 @@ Feature: Patient Search
...
-    When the user search for "<patientName>"
-    Then the result should be "<result>"
-    Examples:
-      | patientName | result           |
-      | Kevin Jones | Found 1 patient  |
-      | 100MQ       | No results found |
+  @patient-search
+  Scenario: Search for a patient
+    When the user search for "John Doe"
+    Then the result should be "Found 1 patient"
```
Comment (jayasanka-sack) [inline qaframework-bdd-tests/src/test/resources/features/refapp-3.x/06-patient-search/patient-search.feature:8]:
```suggestion
    Then the result should be "Found 1 patient"

```

**Review by brandones — APPROVED** [review]
Awesome! So glad to see this done. Great work @kumuditha-udayanga .

Let's give @ibacher a chance to take a look too, since he has a better idea how these docker images are supposed to work.

EDIT: My approval also refers to the changes requested by @jayasanka-sack , which should be made.

**Review by ibacher — APPROVED** [review]
LGTM. Thanks @kumuditha-udayanga!

File: qaframework-bdd-tests/docker/docker-compose-refqa-3x.yml [hunk 5]
```diff
@@ -0,0 +1,57 @@
+version: "3.7"
+
+services:
+  # Proxy
+  proxy:
+    image: openmrs/openmrs-reference-application-3-gateway:latest
+    depends_on:
+      - frontend
+      - openmrs
```
Comment (dkayiwa) [inline qaframework-bdd-tests/docker/docker-compose-refqa-3x.yml:9]: Wouldn't it be better to rename this from `openmrs` to `backend`?

**Comment by jayasanka-sack** [comment]
@ibacher, @brandones  quick question,
Where can I find the source of the following images?

1. openmrs/openmrs-reference-application-3-gateway
2. openmrs/openmrs-reference-application-3-frontend

**Review by sherrif10 — COMMENTED** [review]
@jayasanka-sack Is this going to be an independent run for only Patient Search or the entire ref 3.x workflows. if not i guess we might change the title of the pull request to reflect the work attended to achieve.

**Comment by sherrif10** [comment]
@kumuditha-udayanga  .One more question is. am also trying to achieve this [here](https://github.com/openmrs/openmrs-contrib-qaframework/pull/328). so probably we can pick some similarities from there. 

Is this dockerizing of tests going to be using default refapp 3.x docker image. Can we generic it to use the current  docker image of chrome/firefox to spin up the tests.One advantage this brings is that, we will be no longer depend on the default configured chrome /firefox drivers locally but using the standalone chrome images would suffice. cc @jayasanka-sack  @dkayiwa @ibacher @brandones . Your thoughts about my opinion.

File: qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test.properties [hunk 6]
```diff
@@ -1,7 +1,7 @@
-webapp.url=https://qa-refapp.openmrs.org/openmrs
+webapp.url=http://localhost/openmrs
```
Comment (kdaud) [inline qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test.properties:1]: isn't test.properties file settings responsible for 2.x  RefApp  workflow tests?

Comment (jayasanka-sack) [reply qaframework-bdd-tests/src/test/resources/org/openmrs/uitestframework/test.properties:1]: Oh yes, please remove this @kumuditha-udayanga. 

Thanks @kdaud !

File: qaframework-bdd-tests/docker/docker-compose-refqa-3x.yml (hunk 5 above)
Comment (jayasanka-sack) [reply qaframework-bdd-tests/docker/docker-compose-refqa-3x.yml:9]: That would be nice! <3

**Comment by kdaud** [comment]
@kumuditha-udayanga have you had a chance to take a look at this? https://www.freecodecamp.org/news/writing-good-commit-messages-a-practical-guide/

**Comment by kumuditha-udayanga** [comment]
> @kumuditha-udayanga have you had a chance to take a look at this? https://www.freecodecamp.org/news/writing-good-commit-messages-a-practical-guide/

**Comment by kumuditha-udayanga** [comment]
Thank you! @kdaud, I'll take a look at it.

File: qaframework-bdd-tests/cypress/support/commands.js [hunk 7]
```diff
@@ -333,7 +333,7 @@ Cypress.Commands.add('createPatient', (identifier = null) => {
             {
                 identifier,
                 identifierType: '05a29f94-c0ed-11e2-94be-8c13b969e334',
-                location: '6351fcf4-e311-4a19-90f9-35667d99a8af',
+                location: 'ba685651-ed3b-4e63-9b35-78893060758a',
```
Comment (Piumal1999) [inline qaframework-bdd-tests/cypress/support/commands.js:336]:
```suggestion
                location: DEFAULT_LOCATION_UUID,
```
We can get the location the default location uuid global variable

File: qaframework-bdd-tests/cypress/support/commands.js [hunk 8]
```diff
@@ -76,14 +76,14 @@ Cypress.Commands.add('generateLabResults', (uuid) => {
     const labResults = {
         patient: uuid,
         encounterDatetime: "2012-02-05T04:03:02",
-        location: '6351fcf4-e311-4a19-90f9-35667d99a8af',
+        location: 'ba685651-ed3b-4e63-9b35-78893060758a',
```
Comment (Piumal1999) [inline qaframework-bdd-tests/cypress/support/commands.js:79]:
```suggestion
        location: DEFAULT_LOCATION_UUID,
```
Here also we can use the global variable

**Review by Piumal1999 — COMMENTED** [review]
@kumuditha-udayanga Added 2 minor suggestions

**Comment by ibacher** [comment]
> @ibacher, @brandones quick question, Where can I find sources of the following images?
> 
>     1. openmrs/openmrs-reference-application-3-gateway

https://github.com/openmrs/openmrs-distro-referenceapplication/tree/3.x/run/docker/proxy

>     2. openmrs/openmrs-reference-application-3-frontend

https://github.com/openmrs/openmrs-distro-referenceapplication/tree/3.x/run/docker/ui


> Is this dockerizing of tests going to be using default refapp 3.x docker image. Can we generic it to use the current docker image of chrome/firefox to spin up the tests.One advantage this brings is that, we will be no longer depend on the default configured chrome /firefox drivers locally but using the standalone chrome images would suffice. cc @jayasanka-sack @dkayiwa @ibacher @brandones . Your thoughts about my opinion.

3.x E2E tests use [Cypress](https://www.cypress.io/) which includes it's own browser (essentially a packaged version of Chrome) so this is a non-issue for these tests. While Cypress can be run from a Docker image, it's more flexible if it's installed on the machine locally (i.e., it's quite complex to run Cypress with a UI on Windows and Mac from a Docker image–and the UI is useful when developing and debugging tests).

File: qaframework-bdd-tests/docker/docker-compose-refqa-3x.yml (hunk 5 above)
Comment (ibacher) [reply qaframework-bdd-tests/docker/docker-compose-refqa-3x.yml:9]: So, the proxy's configuration unfortunately [hard-codes](https://github.com/openmrs/openmrs-distro-referenceapplication/blob/6fa42a5dd8071b2726368bba91d854dc15e9a91b/run/docker/proxy/nginx.conf#L39) the name `openmrs`. I think you might be able to get away with:

```yml
  proxy:
    image: openmrs/openmrs-reference-application-3-gateway:latest
    depends_on:
      - frontend
      - backend
    links:
      - backend:openmrs
    ports:
      - "80:80"

# ...

  backend:
    image: openmrs/openmrs-reference-application-3-backend:latest
    depends_on:
      - db
    environment:
      OMRS_CONFIG_MODULE_WEB_ADMIN: "true"
      OMRS_CONFIG_AUTO_UPDATE_DATABASE: "true"
      OMRS_CONFIG_CREATE_TABLES: "true"
      OMRS_CONFIG_CONNECTION_SERVER: db
      OMRS_CONFIG_CONNECTION_DATABASE: openmrs
      OMRS_CONFIG_CONNECTION_USERNAME: openmrs
      OMRS_CONFIG_CONNECTION_PASSWORD: Admin123
    healthcheck:
      test: ["CMD", "curl", "-f", "http://localhost:8080/openmrs"]
      timeout: 5s
    volumes:
      - openmrs-data:/openmrs/data
```

Or a custom nginx.conf file [like we do for dev3.openmrs.org](https://github.com/openmrs/openmrs-contrib-ansible-docker-compose/tree/master/files/emr-3-dev).

Of course, I suppose we could just change the image too...

Comment (brandones) [reply qaframework-bdd-tests/docker/docker-compose-refqa-3x.yml:9]: I don't think it's worth adding any complexity for this.

**Comment by jayasanka-sack** [comment]
Thanks all for your reviews! 🤗 I've created a new branch named `dockerize-3.x`. @kumuditha-udayanga Could you please change point this PR to that branch? So that we can merge once all the tests are migrated to run with this setup.

**Comment by kumuditha-udayanga** [comment]
Sure!

**Review by jayasanka-sack — APPROVED** [review]
Thank you very much @kumuditha-udayanga !!! 🎉🤗

**Pull request merged**